Add support for getting components by name to `struct_exprt`

thomasspriggs opened the pull request:
Add support for getting components by name to `struct_exprt`

thomasspriggs committed:
Add support for getting components by name to `struct_exprt`

This commit uses a template to implement both the const and non-const
overloads of `struct_exprt::component`. This reduces the overhead of
maintaining 2 identical versions and means that both overloads must be
implemented in the same way.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -1823,6 +1823,18 @@ class struct_exprt:public exprt
+  /// \return The expression for a named component of this struct.
+  exprt &component(const irep_idt name)
+  {
+    return operands()[to_struct_type(type()).component_number(name)];
```
Please do
````c++
const auto index = to_struct_type(type()).component_number(name);
DATA_INVARIANT(index < operands().size(), "component matching index should exist");
return operands()[index];
````

thomasspriggs replied:
Done.

tautschnig commented on src/util/std_expr.h:
```diff
@@ -1823,6 +1823,18 @@ class struct_exprt:public exprt
   }
+
+  /// \return The expression for a named component of this struct.
+  exprt &component(const irep_idt name)
```
Pass as a reference

thomasspriggs replied:
Done.

tautschnig reviewed (changes requested):
Good idea, but more defensiveness required as outlined below.

tautschnig reviewed (approved):
Thanks a lot for the immediate updates, I'd suggest squashing before the merge.

thomasspriggs commented:
@tautschnig Will do. My responses to comments are in separate commits for convenience of reviewing only.

tautschnig commented:
> My responses to comments are in separate commits for convenience of reviewing only.

Yes, thank you, I should have explicitly ack'ed that. My note was mainly aimed at anyone else overly eager to merge this, which I hadn't quite made clear.

kroening reviewed (changes requested):
This won't work.
We don't want the struct_type to be the type of the struct_expr (which can be huge, esp. in OO languages), but we want the struct_tag type (which is constant size).
Thus, I recommend passing a namespace, and then using .follow() to get the struct_type.

thomasspriggs commented:
@kroening I have added a commit which uses `ns.follow` in order to address your comment. Please can you re-review?

I have also added a commit which uses a template to de-duplicate the const and non-const overloads of `struct_exprt::component`, as there was slightly more code duplication than I was happy with. However if there is a view that the duplication is better than adding template usage, I am ok with dropping the extra commit.

My intention is to squash all the commits together before this PR is merged. The commits are currently separate for ease of reviewing.

kroening commented:
A rebase should fix the problem with AWS CodeBuild.
